xdsl/folder.py

```diff
@@ -0,0 +1,80 @@
+from collections.abc import Sequence
+
+from xdsl.builder import Builder
+from xdsl.context import Context
+from xdsl.dialect_interfaces import ConstantMaterializationInterface
+from xdsl.interfaces import HasFolderInterface
+from xdsl.ir import Attribute, Operation, SSAValue, TypeAttribute
+from xdsl.rewriter import Rewriter
+
+
+class Folder:
+    context: Context
+
+    def try_fold(self, op: Operation) -> tuple[list[SSAValue], list[Operation]] | None:
+        """
+        Try to fold the given operation.
+        Returns a tuple the list of SSAValues that replace the results of the operation,
+        and a list of new operations that were created during folding.
+        If the operation could not be folded, returns None.
+        """
+
+        if not isinstance(op, HasFolderInterface):
+            return None
+        folded = op.fold()
+        if folded is None:
+            return None
+        results: list[SSAValue] = []
+        new_ops: list[Operation] = []
+        for val, original_result in zip(folded, op.results):
+            if isinstance(val, SSAValue):
+                results.append(val)
+            else:
+                assert isinstance(val, Attribute)
+                dialect = self.context.get_dialect(op.dialect_name())
+                interface = dialect.get_interface(ConstantMaterializationInterface)
+                if not interface:
+                    return None
+                assert isinstance(type := original_result.type, TypeAttribute)
+                new_op = interface.materialize_constant(val, type)
+                if new_op is None:
+                    return None
+                new_ops.append(new_op)
+                results.append(new_op.results[0])
+        return results, new_ops
+
+    def insert_with_fold(
+        self, op: Operation, builder: Builder
+    ) -> tuple[Sequence[SSAValue], list[Operation]] | None:
+        """
+        Inserts the operation using the provided builder, trying to fold it first.
+
+        If folding is successful, returns a tuple containing
+        the folded results and any newly created operations.
+        Otherwise, returns None.
+        """
+        results = self.try_fold(op)
+        if results is not None:
+            values, new_ops = results
+            builder.insert_op(new_ops)
+            return values, new_ops
+        else:
+            builder.insert(op)
+            return op.results, list[Operation]()
+
+    def replace_with_fold(
+        self, op: Operation, safe_erase: bool = True
+    ) -> tuple[list[SSAValue], list[Operation]] | None:
```

Comment:
Feels a little weird not to pass a PatternRewriter here, it might be important for the listener to work